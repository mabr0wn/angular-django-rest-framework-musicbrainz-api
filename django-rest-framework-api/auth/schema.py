# Django
from django.contrib.auth.models import User
# Grphene
import graphene
from graphene_django.types import DjangoObjectType

class CurrentUserNode(DjangoObjectType):

    class Meta:
        model = User
        interfaces = (graphene.relay.Node, )

class Query(object):
    current_user = graphene.Field(CurrentUserNode)

    def resolve_current_user(self, info):
        if not info.context.user.is_authenticated:
            return None
        return info.context.user
