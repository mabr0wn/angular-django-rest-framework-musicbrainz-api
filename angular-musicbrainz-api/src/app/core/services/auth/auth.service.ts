// Angular
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Router } from '@angular/router';
// RxJS
import { Observable, Subject} from 'rxjs';
import { tap } from 'rxjs/operators';

import { CredentialsService, Credentials } from '@core/services/auth/credentials.service';

const httpOptions = {
  headers: new HttpHeaders({ 'Content-Type': 'application/json' })
};

const backend = 'http://localhost:8000/api-auth/'

@Injectable({
  providedIn: 'root'
})
export class AuthService {
  authToken: string;
  usernameSub: Subject<string> = new Subject<string>();
  usernameStr: string;
  isLoggedIn: boolean = false;

  constructor(
    private credentialsService: CredentialsService,
    private http: HttpClient, 
    private router: Router
    ) { }

  loadCreds() {
    let creds = this.credentialsService.credentials
    if (creds) {
      this.setUserName(creds.username);
      this.authToken = creds.token;
    }
  }

  logIn(userInfo): Observable<any> {
    const data = {
      username: userInfo.username,
      password: userInfo.password
    }
    return this.http.post<Credentials>(backend + 'login/', data, httpOptions)
      .pipe( 
        tap(res => {
          this.authToken = res.token;
          this.setUserName(userInfo.username);
          this.credentialsService.setCredentials()
          // if (userInfo.remember) {
          //   localStorage.setItem('userCreds', JSON.stringify({
          //     username: userInfo.username,
          //     token: res.token
          //   }));
          // }
          this.isLoggedIn = true;
        })
      )
  }

  getUsername() {

  }

  sendUsername() {

  }

  private setUserName(name: string) {
    this.usernameStr = name;
    this.usernameSub.next(name);
  }

  getAuthToken() {

  }

  logOut() {

  }

}
