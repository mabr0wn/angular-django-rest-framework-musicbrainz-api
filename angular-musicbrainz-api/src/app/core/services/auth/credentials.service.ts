import { Injectable } from '@angular/core';

export interface Credentials {
  username: string;
  token: string;
}

const credentialKey = 'credentials';

@Injectable({
  providedIn: 'root'
})
export class CredentialsService {

  private _credentials: Credentials| null = null;

  constructor() { 
    const savedCredentials = sessionStorage.getItem(credentialKey) || localStorage.getItem(credentialKey);
    if (savedCredentials) {
      this._credentials = JSON.parse(savedCredentials);
    }
  }

  isAuthenticated(): boolean {
    return !!this.credentials
  }

  get credentials() {
    return this._credentials || null;
  }

  setCredentials(credentials?: Credentials, remember?: boolean) {
    this._credentials = credentials || null;

    if (credentials) {
      const storage  = remember ? localStorage : sessionStorage;
      storage.setItem(credentialKey, JSON.stringify(credentials));
    } else {
      sessionStorage.removeItem(credentialKey);
      localStorage.removeItem(credentialKey);
    }
  }
}


