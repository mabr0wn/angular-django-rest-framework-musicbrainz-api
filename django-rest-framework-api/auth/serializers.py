# Django
from django.contrib.auth.models import User
# Django rest
from rest_framework import serializers



"""
Now we need to add a User serializer to work with Users, we will add a representation of those users
to our API.  We create this to see how many JSON each user has created, it will display the id, username
and the number of Artist that user created.

Because 'artists' is a reverse relationship on the User model, it will not be included by default
when using the ModelSerializer class, so we need to add an explicit field for it.
"""

class UserSerializer(serializers.HyperlinkedModelSerializer):
#   artists = serializers.PrimaryKeyRelatedField(many=True, queryset=Experiment.objects.all())
  
  class Meta:
      model = User
      fields = ('url', 'id', 'username')