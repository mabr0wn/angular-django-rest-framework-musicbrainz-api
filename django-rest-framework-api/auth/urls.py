# Django
from django.urls import path, include
# Django rest
from rest_framework import routers
# Local
from auth import views

"""
Create a router and register our viewsets with it.
DefaultRouter() automatically created the api_root
from views, no longer required.
"""
router = routers.SimpleRouter()
router.register(r'users', views.UserViewSet)


"""
The API URLs are now determined automatically by the router.
Additionally, we include the login URLs for the browsable API.
"""
urlpatterns = [
	# api
    path(r'api/', include(router.urls)),
	path(r'api-auth/', include('rest_framework.urls', namespace='rest_framework')),
]