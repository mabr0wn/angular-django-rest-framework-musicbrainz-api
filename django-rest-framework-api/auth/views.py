from django.shortcuts import render
from auth.serializers import UserSerializer
from django.contrib.auth.models import User


# Create your views here.

class UserViewSet(viewsets.ReadOnlyModelViewSet):
    """
    This Viewset automatically provides 'list' and 'detail' actions
    using the ReadOnly model we know this will provide reas-only
    """
    queryset = User.objects.all()
    serializer_class = UserSerializer