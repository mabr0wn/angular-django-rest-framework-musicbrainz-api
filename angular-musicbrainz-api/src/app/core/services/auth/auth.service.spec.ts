import {
  fakeAsync,
  TestBed,
  tick
} from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { HttpClient} from '@angular/common/http';
// RxJS
import { of } from 'rxjs';
import { AuthService } from './auth.service';
// Credentials
import { CredentialsService, Credentials } from './credentials.service';
import { MockCredentialsService } from './mocks/credentials.service.mock';


describe('AuthService', () => {
  let authService: AuthService;
  let credentialsService: MockCredentialsService;

  const http = {
    post: jest.fn()
  };

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        RouterTestingModule
      ],
      providers: [
        AuthService,
        { provide: CredentialsService, useClass: MockCredentialsService },
        { provide: HttpClient, useValue: http }
      ]
    });

    authService = TestBed.get(AuthService);
    credentialsService = TestBed.get(CredentialsService);
    credentialsService.credentials = null;

  });

  describe('Login', () => {
    test('Should mock http post for user login', fakeAsync(() => {
      // nothing should post yet
      expect(http.post).not.toHaveBeenCalled();
      // mocked data
      const data = {
        username: 'cheeta',
        password: '1234'
      };
      // wait...
      tick();
      // mock url
      let argUrl: string;
      // mock url post
      http.post.mockImplementation(
        (url) => { argUrl = url; return of([]);
      });
      // subscribe to mocked data...
      authService.login(data).subscribe();
      // expect post to been called one time.
      expect(http.post).toHaveBeenCalledTimes(1);
      // mock url should contain login...
      expect(argUrl).toContain('login/');
    }));
    test('Should Authenticate user', fakeAsync(() => {
      expect(credentialsService.isAuthenticated()).toBe(false);
      // Mock
      const data = {
        username: 'cheeta',
        password: '1234',
      };
      tick();

      authService.login(data).subscribe();

      expect(credentialsService.isAuthenticated()).toBe(true);
      expect(credentialsService.credentials).not.toBeNull();
    }));
    test('Should obtain credentials for the session', fakeAsync(() => {
      const spy = jest.spyOn(credentialsService, 'setCredentials');
      const isAuthenticated = credentialsService.setCredentials;
      // Mock
      const data = {
        username: 'cheeta',
        password: '1234'
      };
      tick();

      authService.login(data).subscribe();

      expect(isAuthenticated).toHaveBeenCalled();
      expect(spy.mock.calls[0][1]).toBe(undefined);
    }));
    test('should obtain and remember credentials across session', fakeAsync(() => {
      const spy = jest.spyOn(credentialsService, 'setCredentials');
      const isAuthenticated = credentialsService.setCredentials;
      // Mock
      const data = {
        username: 'cheeta',
        password: '1234',
        remember: true
      };
      tick();

      authService.login(data).subscribe();

      expect(isAuthenticated).toHaveBeenCalled();
      expect(spy.mock.calls[0][1]).toBe(true);
    }));
    test('should remove user authentication', fakeAsync(() => {

      // Mock
      const data = {
        username: 'cheeta',
        password: '1234',
      };
      tick();

      authService.login(data).subscribe(() => {
        expect(credentialsService.isAuthenticated()).toBe(true);

        const request = authService.logout();
        tick();

        request.subscribe(() => {
          expect(credentialsService.isAuthenticated()).toBe(false);
          expect(credentialsService.credentials).toBeNull();
        });
      });
    }));
  });
});
