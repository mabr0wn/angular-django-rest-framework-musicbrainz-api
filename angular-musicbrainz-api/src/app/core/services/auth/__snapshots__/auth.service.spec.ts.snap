// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthService Login Should Authenticate user 1`] = `false`;

exports[`AuthService Login Should Authenticate user 2`] = `true`;

exports[`AuthService Login Should mock http post for user login: login/ 1`] = `"http://localhost:8000/api-auth/login/"`;

exports[`AuthService Login Should obtain credentials for the session 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [],
      undefined,
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`AuthService Login Should obtain credentials for the session 2`] = `undefined`;

exports[`AuthService Login should obtain and remember credentials across session 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [],
      true,
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`AuthService Login should obtain and remember credentials across session 2`] = `true`;
